lutron-config.js

```diff
@@ -86,6 +128,16 @@ module.exports = function (RED) {
                 }
             }
         }
+
+        // Cleanup on close
+        node.on('close', function(done) {
+            node.log('Node shutting down');
+            node.connected = false;
+            node.telnet.end()
+                .then(() => done())
+                .catch(() => this.telnet.destroy().then(() => done()));
```

Comment:
Won't this run `done()` a second time if the `end()` succeeds and the first `done()` throws an exception?

Reply:
No, done shouldn't throw an exception since it's a callback provided by nodered. The catch is in case the end() throws an exception to ensure we still force the cleanup and call done().

Reply:
It shouldn't throw, but you don't control nodered's code. This would be safer and simpler to write as

```suggestion
            node.telnet.end()
                .catch(() => this.telnet.destroy())
                .then(() => done());
```

This way an exception from `done()` won't cause it to repeat (or to call `destroy()`), and you're not repeating the callback either.

Also, question: What happens if `destroy()` throws an exception? Neither version of the code will call `done()` in that case. Maybe it should? If it should, then my version would use `.finally(() => done())` instead (or just `.finally(done)` as `finally` already passes no arguments).

Reply:
Ah, forgot finally was available with promises. Thanks, I'll update.